scripts/delete-old-app-engine-versions.sh

```diff
@@ -0,0 +1,181 @@
+#!/bin/bash
+#
+# Deletes old Google App Engine deployments of Terra UI in an environment. You
+# MUST have GNU date installed to be able to use this script.
+#
+# USAGE: ./delete-old-app-engine-versions.sh ENV
+#   ENV must be one of dev, alpha, staging, or perf
+#
+
+set -eu
+set -o pipefail
+
+# check if colors are supported in the terminal
+check_color_support() {
+    NCOLORS=$(tput colors)
+    if [ "$NCOLORS" -ge 8 ]; then
+        BLD="$(tput bold)"
+        RED="$(tput setaf 1)"
+        GRN="$(tput setaf 2)"
+        RST="$(tput sgr0)"
+    else
+        BLD=""
+        RED=""
+        GRN=""
+        RST=""
+    fi
+    INFO="${BLD}+${RST}"
+}
+
+# print out usage to stdout
+usage() {
+    printf "Usage: %s ${BLD}ENV${RST}\n  ${BLD}ENV${RST} must be one of dev, alpha, staging, or perf\n" "$0"
+    exit 0
+}
+
+# print out error with help message to stderr and exit
+error() {
+    printf "${RED}ERROR: %s${RST}\n\nTry ${BLD}%s --help${RST} to see a list of all options.\n" "$1" "$0" >&2
+    exit 1
+}
+
+# print out error to stderr and exit
+abort() {
+    printf "${RED}ABORT: %s${RST}\n" "$1" >&2
+    exit 1
+}
+
+# always reset gcloud project regardless of error or exit
+reset() {
+    gcloud config set project "${OLD_PROJECT}" 2>/dev/null
+}
+
+# ensure that date from GNU coreutils is installed
+check_gnu_date_installed() {
+    if ! date --version 1>/dev/null 2>&1; then
+        abort "date from GNU coreutils is required"
+    fi
+}
+
+# switch gcloud project to appropriate terra ui environment
+switch_gcloud_project() {
+    gcloud config set project "${NEW_PROJECT}" 2>/dev/null
+    printf "${INFO} Switched the project to ${GRN}%s${RST}\n" "${NEW_PROJECT}"
+}
+
+# ensure that user has appropriate permissions for app engine
+check_user_permissions() {
+    if ! gcloud app services list 1>/dev/null 2>&1; then
```

Comment:
Does this actually check necessary permissions for this script? A user could have viewer access but not be able to delete versions.

Reply:
That's true. Do you have a suggestion for what we should try instead? I don't see a good way of doing this without parsing the error from the `gcloud app versions delete` command.

Reply:
The approach I would take is not checking for permissions in advance and letting the `gcloud app versions delete` command fail if the user doesn't have adequate permissions. The error message from gcloud is likely informative enough for the audience using this script.

https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/ (the article is about Python, but the principle applies)

Reply:
I am still trying to identify the best way to check that the user has the right permissions to delete without having to parse the error.